From af692594530489794e5bfc2782dcfc823318c5d4 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Mon, 19 May 2025 09:43:33 +0200
Subject: [PATCH] fix 32bit response size limit

the response size was an int which limits the response size to 32bit. But sometimes
the results get larger. Change the type into a size_t which is the correct type
for this.
---
 src/AndingFilter.cc |  2 +-
 src/AndingFilter.h  |  2 +-
 src/LogEntry.cc     |  6 +++---
 src/LogEntry.h      |  8 +++++---
 src/Logfile.cc      | 12 ++++++------
 src/Logfile.h       | 12 ++++++------
 src/OutputBuffer.cc | 20 ++++++++++----------
 src/OutputBuffer.h  | 10 +++++-----
 src/Query.cc        | 26 +++++++++++++-------------
 src/Store.cc        |  2 +-
 src/module.c        |  8 ++++----
 src/unixcat.c       |  6 +++---
 12 files changed, 58 insertions(+), 56 deletions(-)

diff --git a/src/AndingFilter.cc b/src/AndingFilter.cc
index 4255a73..1617ff0 100644
--- a/src/AndingFilter.cc
+++ b/src/AndingFilter.cc
@@ -23,7 +23,7 @@ void AndingFilter::addSubfilter(Filter *f)
 }
 
 
-Filter *AndingFilter::stealLastSubfiler()
+Filter *AndingFilter::stealLastSubfilter()
 {
     if (_subfilters.size() == 0)
         return 0;
diff --git a/src/AndingFilter.h b/src/AndingFilter.h
index 4385988..083cea4 100644
--- a/src/AndingFilter.h
+++ b/src/AndingFilter.h
@@ -19,7 +19,7 @@ class AndingFilter : public Filter
     AndingFilter() {}
     ~AndingFilter();
     void addSubfilter(Filter *);
-    Filter *stealLastSubfiler();
+    Filter *stealLastSubfilter();
     bool accepts(void *data);
     void combineFilters(int count, int andor);
     unsigned numFilters() { return _subfilters.size(); }
diff --git a/src/LogEntry.cc b/src/LogEntry.cc
index d835ed8..a1d9f59 100644
--- a/src/LogEntry.cc
+++ b/src/LogEntry.cc
@@ -10,7 +10,7 @@
 #include "strutil.h"
 #include "logger.h"
 
-LogEntry::LogEntry(unsigned lineno, char *line)
+LogEntry::LogEntry(uint64_t lineno, char *line)
 {
     // zero all elements as fast as possible -> default values
     bzero(this, sizeof(LogEntry));
@@ -465,9 +465,9 @@ inline int LogEntry::hostStateToInt(char *s)
     }
 }
 
-unsigned LogEntry::updateReferences()
+uint64_t LogEntry::updateReferences()
 {
-    unsigned updated = 0;
+    uint64_t updated = 0;
     if (_host_name) {
         _host = find_host(_host_name);
         updated++;
diff --git a/src/LogEntry.h b/src/LogEntry.h
index 4ea35e4..6268021 100644
--- a/src/LogEntry.h
+++ b/src/LogEntry.h
@@ -5,6 +5,8 @@
 #ifndef LogEntry_h
 #define LogEntry_h
 
+#include <stdint.h>
+
 #define LOGCLASS_INFO              0 // all messages not in any other class
 #define LOGCLASS_ALERT             1 // alerts: the change service/host state
 #define LOGCLASS_PROGRAM           2 // important programm events (restart, ...)
@@ -41,7 +43,7 @@ enum LogEntryType {
 
 struct LogEntry
 {
-    unsigned     _lineno;      // line number in file
+    uint64_t     _lineno;      // line number in file
     time_t       _time;
     unsigned     _logclass;
     LogEntryType _type;
@@ -65,9 +67,9 @@ struct LogEntry
     contact     *_contact;
     command     *_command;
 
-    LogEntry(unsigned lineno, char *line);
+    LogEntry(uint64_t lineno, char *line);
     ~LogEntry();
-    unsigned updateReferences();
+    uint64_t updateReferences();
     static int serviceStateToInt(char *s);
     static int hostStateToInt(char *s);
 
diff --git a/src/Logfile.cc b/src/Logfile.cc
index 1c81680..60f32f2 100644
--- a/src/Logfile.cc
+++ b/src/Logfile.cc
@@ -166,7 +166,7 @@ long Logfile::freeMessages(unsigned logclasses)
     return freed;
 }
 
-inline bool Logfile::processLogLine(uint32_t lineno, unsigned logclasses)
+inline bool Logfile::processLogLine(uint64_t lineno, unsigned logclasses)
 {
     LogEntry *entry = new LogEntry(lineno, _linebuffer);
     // ignored invalid lines
@@ -241,7 +241,7 @@ uint64_t Logfile::makeKey(time_t t, unsigned lineno)
 // to a malloced buffer, that the caller must free (or 0, in case of
 // an error). The buffer is 2 bytes larger then the file. One byte
 // at the beginning and at the end of the buffer are '\0'.
-char *Logfile::readIntoBuffer(int *size)
+char *Logfile::readIntoBuffer(size_t *size)
 {
     int fd = open(_path, O_RDONLY);
     if (fd < 0) {
@@ -266,15 +266,15 @@ char *Logfile::readIntoBuffer(int *size)
         return 0;
     }
 
-    int r = read(fd, buffer + 1, *size);
+    size_t r = read(fd, buffer + 1, *size);
     if (r < 0) {
-        logger(LG_WARN, "Cannot read %d bytes from %s: %s", *size, _path, strerror(errno));
+        logger(LG_WARN, "Cannot read %zu bytes from %s: %s", *size, _path, strerror(errno));
         free(buffer);
         close(fd);
         return 0;
     }
     else if (r != *size) {
-        logger(LG_WARN, "Read only %d out of %d bytes from %s", r, *size, _path);
+        logger(LG_WARN, "Read only %zu out of %zu bytes from %s", r, *size, _path);
         free(buffer);
         close(fd);
         return 0;
@@ -335,4 +335,4 @@ time_t Logfile::end()
 
     close(fd);
     return _end;
-}
\ No newline at end of file
+}
diff --git a/src/Logfile.h b/src/Logfile.h
index b4f7ec4..a744888 100644
--- a/src/Logfile.h
+++ b/src/Logfile.h
@@ -29,7 +29,7 @@ class Logfile
     bool       _watch;         // true only for current logfile
     ino_t      _inode;         // needed to detect switching
     fpos_t     _read_pos;      // read until this position
-    uint32_t   _lineno;        // read until this line
+    uint64_t   _lineno;        // read until this line
 
     logfile_entries_t  _entries;
     char       _linebuffer[MAX_LOGLINE];
@@ -40,7 +40,7 @@ class Logfile
     ~Logfile();
 
     char *path() { return _path; }
-    char *readIntoBuffer(int *size);
+    char *readIntoBuffer(size_t *size);
     void load(LogCache *LogCache, time_t since, time_t until, unsigned logclasses);
     void flush();
     time_t since() { return _since; }
@@ -48,9 +48,9 @@ class Logfile
     bool watch() { return _watch; }
     unsigned classesRead() { return _logclasses_read; }
     long numEntries() { return _entries.size(); }
-    logfile_entries_t* getEntriesFromQuery(Query *query, LogCache *lc, time_t since, time_t until, unsigned);
-    bool answerQuery(Query *query, LogCache *lc, time_t since, time_t until, unsigned);
-    bool answerQueryReverse(Query *query, LogCache *lc, time_t since, time_t until, unsigned);
+    logfile_entries_t* getEntriesFromQuery(Query *query, LogCache *lc, time_t since, time_t until, unsigned logclasses);
+    bool answerQuery(Query *query, LogCache *lc, time_t since, time_t until, unsigned logclasses);
+    bool answerQueryReverse(Query *query, LogCache *lc, time_t since, time_t until, unsigned logclasses);
 
     long freeMessages(unsigned logclasses);
 
@@ -60,7 +60,7 @@ class Logfile
 private:
     void loadRange(FILE *file, unsigned missing_types, LogCache *,
                    time_t since, time_t until, unsigned logclasses);
-    bool processLogLine(uint32_t, unsigned);
+    bool processLogLine(uint64_t, unsigned);
     uint64_t makeKey(time_t, unsigned);
 };
 
diff --git a/src/OutputBuffer.cc b/src/OutputBuffer.cc
index d198287..d5cd584 100644
--- a/src/OutputBuffer.cc
+++ b/src/OutputBuffer.cc
@@ -50,22 +50,22 @@ void OutputBuffer::addString(const char *s)
     addBuffer(s, l);
 }
 
-void OutputBuffer::addBuffer(const char *buf, unsigned len)
+void OutputBuffer::addBuffer(const char *buf, size_t len)
 {
     needSpace(len);
     memcpy(_writepos, buf, len);
     _writepos += len;
 }
 
-void OutputBuffer::needSpace(unsigned len)
+void OutputBuffer::needSpace(size_t len)
 {
     if (_writepos + len > _end)
     {
-        unsigned s = size();
-        unsigned needed = s + len;
+        size_t s = size();
+        size_t needed = s + len;
         while (_max_size < needed) {
             // increase until enough space
-            // double untill 500MB
+            // double until 500MB
             if(_max_size < 500 * 1024 * 1024) {
                 _max_size *= 2;
             } else {
@@ -83,7 +83,7 @@ void OutputBuffer::needSpace(unsigned len)
 void OutputBuffer::flush(int fd)
 {
     const char *buffer = _buffer;
-    int s = size();
+    size_t s = size();
     // if response code is not OK, output error
     // message instead of data
     if (_response_code != RESPONSE_CODE_OK)
@@ -94,7 +94,7 @@ void OutputBuffer::flush(int fd)
     if (_response_header == RESPONSE_HEADER_FIXED16)
     {
         char header[17];
-        snprintf(header, sizeof(header), "%03d %11d\n", _response_code, s);
+        snprintf(header, sizeof(header), "%03d %11zu\n", _response_code, s);
         writeData(fd, header, 16);
         writeData(fd, buffer, s);
     }
@@ -104,7 +104,7 @@ void OutputBuffer::flush(int fd)
 }
 
 
-void OutputBuffer::writeData(int fd, const char *write_from, int to_write)
+void OutputBuffer::writeData(int fd, const char *write_from, size_t to_write)
 {
     struct timeval tv;
     while (!*_termination_flag && to_write > 0)
@@ -120,11 +120,11 @@ void OutputBuffer::writeData(int fd, const char *write_from, int to_write)
         if (retval > 0 && FD_ISSET(fd, &fds)) {
             ssize_t w = write(fd, write_from, to_write);
             if (w < 0) {
-                logger(LG_INFO, "Couldn't write %d bytes to client socket: %s", to_write, strerror(errno));
+                logger(LG_INFO, "Couldn't write %zu bytes to client socket: %s", to_write, strerror(errno));
                 break;
             }
             else if (w == 0)
-                logger(LG_INFO, "Strange: wrote 0 bytes inspite of positive select()");
+                logger(LG_INFO, "Strange: wrote 0 bytes in spite of positive select()");
             else {
                 to_write -= w;
             }
diff --git a/src/OutputBuffer.h b/src/OutputBuffer.h
index 30ee3e1..2de9ff0 100644
--- a/src/OutputBuffer.h
+++ b/src/OutputBuffer.h
@@ -27,7 +27,7 @@ class OutputBuffer
     char *_buffer;
     char *_writepos;
     char *_end;
-    unsigned _max_size;
+    size_t _max_size;
     int _response_header;
     unsigned _response_code;
     string _error_message;
@@ -37,10 +37,10 @@ class OutputBuffer
     OutputBuffer(int *termination_flag);
     ~OutputBuffer();
     const char *buffer() { return _buffer; }
-    unsigned size() { return _writepos - _buffer; }
+    size_t size() { return _writepos - _buffer; }
     void addChar(char c);
     void addString(const char *);
-    void addBuffer(const char *, unsigned);
+    void addBuffer(const char *, size_t);
     void reset();
     void flush(int fd);
     bool shouldTerminate();
@@ -53,8 +53,8 @@ class OutputBuffer
     bool isAlive(int fd);
 
 private:
-    void needSpace(unsigned);
-    void writeData(int fd, const char *, int);
+    void needSpace(size_t);
+    void writeData(int fd, const char *, size_t);
 };
 
 
diff --git a/src/Query.cc b/src/Query.cc
index 2d2a570..61434d1 100644
--- a/src/Query.cc
+++ b/src/Query.cc
@@ -292,14 +292,14 @@ void Query::parseNegateLine(char *line, bool filter)
 
     Filter *to_negate;
     if (filter) {
-        to_negate = _filter.stealLastSubfiler();
+        to_negate = _filter.stealLastSubfilter();
         if (!to_negate) {
             _output->setError(RESPONSE_CODE_INVALID_HEADER, "Negate: no Filter: header to negate");
             return;
         }
     }
     else {
-        to_negate = _wait_condition.stealLastSubfiler();
+        to_negate = _wait_condition.stealLastSubfilter();
         if (!to_negate) {
             _output->setError(RESPONSE_CODE_INVALID_HEADER, "Negate: no Wait:-condition negate");
             return;
@@ -784,7 +784,7 @@ void Query::parseLocaltimeLine(char *line)
     time_t our_time = time(0);
 
     // compute offset to be *added* each time we output our time and
-    // *substracted* from reference value by filter headers
+    // *subtracted* from reference value by filter headers
     int dif = their_time - our_time;
 
     // Round difference to half hour. We assume, that both clocks are more
@@ -913,8 +913,8 @@ bool Query::timelimitReached()
 bool Query::processDataset(void *data)
 {
     if (_output->size() > g_max_response_size) {
-        logger(LG_INFO, "Maximum response size of %d bytes exceeded!", g_max_response_size);
-        _output->setError(RESPONSE_CODE_LIMIT_EXCEEDED, "Maximum response size of %d reached", g_max_response_size);
+        logger(LG_INFO, "Maximum response size of %lu bytes exceeded!", g_max_response_size);
+        _output->setError(RESPONSE_CODE_LIMIT_EXCEEDED, "Maximum response size of %lu reached", g_max_response_size);
         return false;
     }
 
@@ -953,13 +953,13 @@ bool Query::processDataset(void *data)
                     _sorter.insert( data, _limit+_offset );
 
                     // make sure we don't create too many aggregation entries. The size is only a rough estimation
-                    // from the last entry mulitplies with the number of entries.
+                    // from the last entry multiplies with the number of entries.
                     size_t rowsize = 0;
                     for (_stats_group_spec_t::iterator iit = groupspec.begin(); iit != groupspec.end(); ++iit)
                         rowsize += sizeof(char*) * strlen((*iit).c_str());
                     if (_sorter.size() * rowsize > g_max_response_size) {
-                        logger(LG_INFO, "Maximum response size of %d bytes exceeded!", g_max_response_size);
-                        _output->setError(RESPONSE_CODE_LIMIT_EXCEEDED, "Maximum response size of %d reached", g_max_response_size);
+                        logger(LG_INFO, "Maximum response size of %lu bytes exceeded!", g_max_response_size);
+                        _output->setError(RESPONSE_CODE_LIMIT_EXCEEDED, "Maximum response size of %lu reached", g_max_response_size);
                         return false;
                     }
                 }
@@ -1063,7 +1063,7 @@ void Query::finish()
             }
             outputDatasetEnd();
         }
-        // output values of all stats groups (output has been post poned until now)
+        // output values of all stats groups (output has been postponed until now)
         for (_stats_groups_t::iterator it = _stats_groups.begin();
                 it != _stats_groups.end();
                 ++it)
@@ -1284,7 +1284,7 @@ void Query::outputString(const char *value)
                 _output->addChar(*r);
             }
 
-            // interprete two-Byte UTF-8 sequences in mode 'utf8' and 'mixed'
+            // interpret two-Byte UTF-8 sequences in mode 'utf8' and 'mixed'
             else if ((g_data_encoding == ENCODING_UTF8 || g_data_encoding == ENCODING_MIXED)
                     && ((*r & 0xE0) == 0xC0)) {
                 outputUnicodeEscape(((*r & 31) << 6) | (*(r+1) & 0x3F)); // 2 byte encoding
@@ -1292,7 +1292,7 @@ void Query::outputString(const char *value)
                 chars_left--;
             }
 
-            // interprete 3/4-Byte UTF-8 sequences only in mode 'utf8'
+            // interpret 3/4-Byte UTF-8 sequences only in mode 'utf8'
             else if (g_data_encoding == ENCODING_UTF8) {
                 // three-byte sequences (avoid buffer overflow!)
                 if ((*r & 0xF0) == 0xE0) {
@@ -1330,7 +1330,7 @@ void Query::outputString(const char *value)
                 }
             }
 
-            // in latin1 and mixed mode interprete all other non-ASCII characters as latin1
+            // in latin1 and mixed mode interpret all other non-ASCII characters as latin1
             else {
                 outputUnicodeEscape((unsigned)((int)*r + 256)); // assume latin1 encoding
             }
@@ -1452,7 +1452,7 @@ void Query::doWait()
             _wait_condition.accepts(_wait_object))
     {
         if (g_debug_level >= 2)
-            logger(LG_INFO, "Wait condition true, no waiting neccessary");
+            logger(LG_INFO, "Wait condition true, no waiting necessary");
         return;
     }
 
diff --git a/src/Store.cc b/src/Store.cc
index 843b994..eb8cfb8 100644
--- a/src/Store.cc
+++ b/src/Store.cc
@@ -203,7 +203,7 @@ void Store::answerGetRequest(InputBuffer *input, OutputBuffer *output, const cha
         gettimeofday(&after, 0);
         unsigned long ustime = (after.tv_sec - before.tv_sec) * 1000000 + (after.tv_usec - before.tv_usec);
         if (g_debug_level > 0)
-            logger(LG_INFO, "Time to process %s request: %lu us. Size of answer: %d bytes", tablename, ustime, output->size());
+            logger(LG_INFO, "Time to process %s request: %lu us. Size of answer: %zu bytes", tablename, ustime, output->size());
     }
 
     if(logcacheLocked)
diff --git a/src/module.c b/src/module.c
index aed4f4e..51f5bae 100644
--- a/src/module.c
+++ b/src/module.c
@@ -98,7 +98,7 @@ static void reap_client_threads()
 
         if ((g_client_threads = realloc(g_client_threads, g_num_client_threads * sizeof(pthread_t))) == NULL) {
             if (g_num_client_threads > 0) {
-                logger(LG_ERR, "Failed to shrink client thread array (Number of threads: %d)", g_num_client_threads);
+                logger(LG_ERR, "Failed to shrink client thread array (Number of threads: %zu)", g_num_client_threads);
             }
             else if (g_debug_level >= 2) {
                 logger(LG_DEBUG, "All client threads reaped.");
@@ -147,7 +147,7 @@ static int accept_connection(int sd, int events, void *discard)
     pthread_attr_destroy(&attr);
 
     if ((g_client_threads = realloc(g_client_threads, (g_num_client_threads + 1) * sizeof(pthread_t))) == NULL) {
-        logger(LG_ERR, "Failed to allocate space for client thread: %s (Number of threads: %d)", strerror(errno), g_num_client_threads + 1);
+        logger(LG_ERR, "Failed to allocate space for client thread: %s (Number of threads: %zu)", strerror(errno), g_num_client_threads + 1);
         close(cc);
         logger(LG_ERR, "Joining with client thread right away to avoid losing it.");
         if (0 != pthread_join(*thr, NULL)) {
@@ -213,7 +213,7 @@ int open_unix_socket()
         return false;
     }
 
-    // Imortant: close on exec -> check plugins must not inherit it!
+    // Important: close on exec -> check plugins must not inherit it!
     if (0 < fcntl(g_socket_fd, F_SETFD, FD_CLOEXEC))
         logger(LG_INFO, "Cannot set FD_CLOEXEC on socket: %s", strerror(errno));
 
@@ -223,7 +223,7 @@ int open_unix_socket()
     strncpy(sockaddr.sun_path, g_socket_addr, sizeof(sockaddr.sun_path));
     if (bind(g_socket_fd, (struct sockaddr *) &sockaddr, SUN_LEN(&sockaddr)) < 0)
     {
-        logger(LG_ERR , "Unable to bind adress %s to UNIX socket: %s",
+        logger(LG_ERR , "Unable to bind address %s to UNIX socket: %s",
                 g_socket_addr, strerror(errno));
         close(g_socket_fd);
         return false;
diff --git a/src/unixcat.c b/src/unixcat.c
index 5fbf38b..007c2d8 100644
--- a/src/unixcat.c
+++ b/src/unixcat.c
@@ -38,7 +38,7 @@ struct thread_info
 };
 
 
-int read_with_timeout(int from, char *buffer, int size, int us)
+size_t read_with_timeout(int from, char *buffer, size_t size, int us)
 {
     fd_set fds;
     FD_ZERO(&fds);
@@ -84,11 +84,11 @@ void *copy_thread(void *info)
         }
         char *write_pos = buffer;
         while (r) {
-            int w = write(to, write_pos, r);
+            size_t w = write(to, write_pos, r);
             if (w > 0)
                 r -= w;
             else if (w == 0 && r > 0) {
-                fprintf(stderr, "Error: Cannot write %d bytes to %d: %s\n", w, to, strerror(errno));
+                fprintf(stderr, "Error: Cannot write %zu bytes to %d: %s\n", w, to, strerror(errno));
                 break;
             }
         }
